### PR TITLE
Validation report view

### PR DESCRIPTION
**Pull request by sgiaccio**

### CODE REVIEW

File: server/modules/record/repository/validationReportRepository.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  ORDER BY r.uuid, h.node_id
+  LIMIT ${limit || 'ALL'}
+  OFFSET ${offset}`
+
+export const fetchValidationReport = async (surveyId, cycle, offset = 0, limit = null, client = db) =>
+  await client.map(query(surveyId, offset, limit), [cycle], res =>
+    Object.entries(res).reduce((obj, [k, v]) => { obj[camelize(k)] = v; return obj }, {}))
+
+export const countValidationReports = async (surveyId, cycle, client = db) =>
+  await client.one(`SELECT count(*) from(${query(surveyId, '0')}) as v`, [cycle])
```
Comment (SteRiccio) [inline server/modules/record/repository/validationReportRepository.js]: the offset parameter is not required. Is the alias "as v" required?

File: webapp/loggedin/modules/data/validationReport/validationReportView.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+  let keysHierarchy = row.keysHierarchy
+  if (keysHierarchy[0].nodeDefUuid === null) {
+    keysHierarchy = keysHierarchy.slice(1)
+  }
+  keysHierarchy = keysHierarchy.concat({ keys: row.keysSelf || {}, nodeDefUuid: row.nodeDefUuid })
+
+  const getKeyValues = keys => Object.values(keys).reduce((values, value) => values.concat(value === null ? '' : value), [])
+
+  const path = keysHierarchy.reduce((p, h) => {
```
Comment (SteRiccio) [inline webapp/loggedin/modules/data/validationReport/validationReportView.js]: "path" is not the final path (String) but it should be actually "pathParts".
Maybe it's better to use in this case a R.pipe and join the parts after the reduce... 
In the reduce iterator function, p and h are not so intuitive... p is the accumulator (acc, accPath or something like this) h is the hiearchyPart, isn't it?

File: webapp/loggedin/modules/data/validationReport/validationReportView.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    const parentNodeDef = Survey.getNodeDefByUuid(h.nodeDefUuid)(survey)
+    const parentNodeDefLabel = NodeDef.getLabel(parentNodeDef, i18n.lang)
+    const keyValues = getKeyValues(h.keys)
```
Comment (SteRiccio) [inline webapp/loggedin/modules/data/validationReport/validationReportView.js]: can you use just a R.reject(R.isNil)(h.keys) ?

File: webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js [hunk 4]
```diff
@@ -63,8 +63,9 @@ const mapStateToProps = (state, props) => {
       if (node) {
         validation = RecordValidation.getNodeValidation(node)(recordValidation)
       } else if (NodeDef.isEntity(nodeDef)) {
-        //only entities can have children with min/max count validation
-        validation = RecordValidation.getValidationChildrenCount(parentNode, nodeDef)(recordValidation)
+        // only entities can have children with min/max count validation
+        validation = Validation.getFieldValidation(parentNode.uuid)(recordValidation)
```
Comment (SteRiccio) [inline webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js]: how can you distinguish validation errors of different node definitions inside the same parent entity now?

File: webapp/commonComponents/validationTooltip.js [hunk 5]
```diff
@@ -1,32 +1,32 @@
...
+    type = 'warning'
+  }
 
-  const messagesHtml = isValid
+  const content = isValid
     ? null
-    : ValidationUtils.getValidationFieldMessagesHTML(i18n, showKeys)(validation)
+    : React.createElement(ValidationFieldMessages, { validation, showKeys })
+
+  const show = Validation.isWarning(validation) || Validation.isError(validation)
```
Comment (SteRiccio) [inline webapp/commonComponents/validationTooltip.js]: can it be just !Validation.isValid(validation)?

File: webapp/commonComponents/validationTooltip.js [hunk 6]
```diff
@@ -1,32 +1,32 @@
...
 
   return (
     <Tooltip
       type={type}
-      messages={messagesHtml}
-      className={className}>
+      content={content}
+      className={className}
+      show={show}>
```
Comment (SteRiccio) [inline webapp/commonComponents/validationTooltip.js]: is it better not to render it at all?

File: webapp/commonComponents/validationFieldMessages.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  // extract invalid fields error messages
+  R.map(([field, fieldValidation]) => {
+    let k = field
+    let v = fieldValidation
+
+    // if it's a childrenCount validation, extract the actual minCount or maxCount validation
+    // and add the nodeDefUuid name as param
+    if (field === RecordValidations.keys.childrenCount) {
+      const [nodeDefUuid, nodeValidation] = Object.entries(Validation.getFieldValidations(v))[0];
+      [k, v] = Object.entries(Validation.getFieldValidations(nodeValidation))[0]
```
Comment (SteRiccio) [inline webapp/commonComponents/validationFieldMessages.js]: you are always considering the first validation in childrenCount... what about the other ones?

File: server/modules/record/repository/validationReportRepository.js (hunk 1 above)
Comment (sgiaccio) [reply server/modules/record/repository/validationReportRepository.js]: Right

File: webapp/commonComponents/validationTooltip.js (hunk 6 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationTooltip.js]: show refers to the tooltip content, renamed it to showContent

File: webapp/loggedin/modules/data/validationReport/validationReportView.js (hunk 2 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.js]: Renamed as you suggested

File: webapp/loggedin/modules/data/validationReport/validationReportView.js (hunk 3 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.js]: Yes

File: core/validation/validation.js [hunk 8]
```diff
@@ -128,10 +128,14 @@ export const getFieldValidation = field => R.pathOr(newInstance(), [keys.fields,
 
 export const getErrors = R.propOr([], keys.errors)
 export const hasErrors = R.pipe(getErrors, R.isEmpty, R.not)
+export const isError = validation =>
+  hasErrors(validation) || R.pipe(getFieldValidations, R.values, R.any(isError))(validation)
```
Comment (minotogna) [inline core/validation/validation.js]: does this work on a nested validation object? (e.g. category) (same for isWarning)

File: webapp/commonComponents/tooltip.js [hunk 9]
```diff
@@ -22,21 +22,16 @@ class Tooltip extends React.Component {
   }
 
   mouseEnter () {
-    const { messages, type = '' } = this.props
-
-    if (!(R.isEmpty(messages) || R.isNil(messages))) {
+    const { messages, content, show, type = '' } = this.props
```
Comment (minotogna) [inline webapp/commonComponents/tooltip.js]: can you add show to defaultProps (and specify in a comment how it's used)

File: webapp/commonComponents/validationTooltip.js [hunk 10]
```diff
@@ -1,32 +1,32 @@
 import React from 'react'
 
-import * as ValidationUtils from '@webapp/utils/validationUtils'
+import ValidationFieldMessages from '@webapp/commonComponents/validationFieldMessages'
```
Comment (minotogna) [inline webapp/commonComponents/validationTooltip.js]: i would add this import where other components are imported

File: webapp/commonComponents/tooltip.js [hunk 11]
```diff
@@ -22,21 +22,16 @@ class Tooltip extends React.Component {
...
 
+    if (show && (content || !(R.isEmpty(messages) || R.isNil(messages)))) {
```
Comment (minotogna) [inline webapp/commonComponents/tooltip.js]: why do you need content props? I would always use messages, and if you are invoking it with just one message (which is your content) you could pass it as ```messages={['my message']}```

File: webapp/commonComponents/validationFieldMessages.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+          <Markdown className="validation_field_message__text" source={message}/>
+        </div>
+      )
+    )(validation)
+
+
+const ValidationFieldMessages = ({ validation, showKeys = true, showIcons = false, survey }) => {
+  const i18n = useI18n()
+
+  return <>{getValidationFieldMessagesHTML(i18n, showKeys, showIcons, survey)(validation)}</>
```
Comment (minotogna) [inline webapp/commonComponents/validationFieldMessages.js]: to make it more clear, I would iterate in here directly

File: webapp/commonComponents/validationFieldMessages.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+const mapStateToProps = state => {
+  const survey = SurveyState.getSurvey(state)
+
+  return {
+    survey,
```
Comment (minotogna) [inline webapp/commonComponents/validationFieldMessages.js]: I wouldn't pass the all survey to the component. you can refactor with inner  connected components

File: webapp/commonComponents/validationFieldMessages.scss [hunk 14]
```diff
@@ -0,0 +1,22 @@
+@import "~@webapp/style/vars";
+
+.validation_field_message {
```
Comment (minotogna) [inline webapp/commonComponents/validationFieldMessages.scss]: we use dash as separator for main css class ```validation-field-message```

File: webapp/commonComponents/validationFieldMessages.scss [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+  &.warning .icon {
+    color: $orange;
+  }
+
+  &.error .icon {
+    color: $red;
+  }
+
+  .validation_field_message__text {
```
Comment (minotogna) [inline webapp/commonComponents/validationFieldMessages.scss]: and this ```validation-field-message__text``` and you can move this outside

File: webapp/loggedin/modules/data/validationReport/validationReportView.js [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+  )
+}
+
+const ValidationReportView = ({ canInvite, user, survey, surveyCycleKey, reloadListItems, history }) => {
+  useOnUpdate(() => {
+    reloadListItems(validationReportModule, { cycle: surveyCycleKey })
+  }, [surveyCycleKey])
+
+  const onRowClick = record => {
+    const parentEntityUuid = R.prop('nodeUuid', R.last(record.keysHierarchy))
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportView.js]: don't we have an object to access the keys? in this case it's NodeKeys?

File: webapp/loggedin/modules/data/validationReport/validationReportView.js [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    <>
+      <div>#</div>
+      <div>{i18n.t('common.path')}</div>
+      <div>{i18n.t('common.messages')}</div>
+      <div />
+    </>
+  )
+}
+
+const ValidationReportRow = ({ user, survey, row, idx, offset }) => {
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportView.js]: I would move this component in its own file

File: webapp/loggedin/modules/data/validationReport/validationReportView.js [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+import * as SurveyState from '@webapp/survey/surveyState'
+import * as RecordsState from '@webapp/loggedin/modules/data/records/recordsState'
+
+import { reloadListItems } from '@webapp/loggedin/tableViews/actions'
+
+import ValidationFieldMessages from '@webapp/commonComponents/validationFieldMessages'
+
+const validationReportModule = 'validationReport'
+
+const ValidationReportRowHeader = ({ nodeDefKeys }) => {
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportView.js]: I would move this component in its own file

File: webapp/loggedin/modules/data/validationReport/validationReportView.scss [hunk 19]
```diff
@@ -0,0 +1,22 @@
+@import "~@webapp/style/vars";
+
+.validation_report_view__table {
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportView.scss]: as above. and i would call this just ```validation-report```

File: webapp/loggedin/modules/data/validationReport/validationReportView.scss [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+  //   background-color: red;
+  // }
+  .table__row {
+    height: unset;
+    // background-color: red;
+    padding: 6px 0;
+  }
+}
+
+.validation_report_view__message {
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportView.scss]: ```validation-report__message```

File: webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js (hunk 4 above)
Comment (minotogna) [reply webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js]: and also, i would use our objects model to get the uuid

File: core/validation/validation.js (hunk 8 above)
Comment (SteRiccio) [reply core/validation/validation.js]: maybe it's better to have 2 functions, one like this, ```isError``` that checkes if there are any errors in the validation object, another one that calls isError recursively, I don't know how to call it, something like ```hasErrorsDeep```

File: webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js (hunk 4 above)
Comment (sgiaccio) [reply webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js]: Done both

File: webapp/commonComponents/tooltip.js (hunk 9 above)
Comment (sgiaccio) [reply webapp/commonComponents/tooltip.js]: Ok

File: webapp/commonComponents/validationTooltip.js (hunk 10 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationTooltip.js]: Ok

File: webapp/loggedin/modules/data/validationReport/validationReportView.js (hunk 17 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.js]: Ok

File: webapp/loggedin/modules/data/validationReport/validationReportView.js (hunk 18 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.js]: Ok

File: webapp/loggedin/modules/data/validationReport/validationReportView.scss (hunk 20 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.scss]: Renamed

File: webapp/loggedin/modules/data/validationReport/validationReportView.scss (hunk 19 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.scss]: Renamed

File: webapp/commonComponents/validationFieldMessages.js (hunk 12 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationFieldMessages.js]: Done

File: webapp/commonComponents/validationFieldMessages.scss (hunk 15 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationFieldMessages.scss]: Ok

File: webapp/commonComponents/validationFieldMessages.scss (hunk 14 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationFieldMessages.scss]: changed

File: webapp/commonComponents/validationFieldMessages.js (hunk 7 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationFieldMessages.js]: Fixed

File: webapp/commonComponents/validationFieldMessages.js (hunk 13 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationFieldMessages.js]: Ok, moved to validationReportRow

File: webapp/loggedin/modules/data/validationReport/validationReportView.js (hunk 16 above)
Comment (sgiaccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.js]: Yes, using NodeKeys

File: webapp/commonComponents/validationTooltip.js (hunk 5 above)
Comment (sgiaccio) [reply webapp/commonComponents/validationTooltip.js]: In some cases you can have isValid = false with no tooltip content

File: webapp/loggedin/modules/data/validationReport/validationReportRow.js [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+      <div>
+        <span className={`icon icon-12px icon-action ${canEdit ? 'icon-pencil2' : 'icon-eye'}`} />
+      </div>
+    </>
+  )
+}
+
+const mapStateToProps = state => ({
+  user: AppState.getUser(state),
+  survey: SurveyState.getSurvey(state),
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportRow.js]: I would not pass the entire survey to the component. You can get the props you need here, in mapStateToProps

File: webapp/loggedin/modules/data/validationReport/validationReportView.js [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+const validationReportModule = 'validationReport'
+
+const ValidationReportView = ({ surveyCycleKey, reloadListItems, history }) => {
+  useOnUpdate(() => {
+    reloadListItems(validationReportModule, { cycle: surveyCycleKey })
+  }, [surveyCycleKey])
+
+  const onRowClick = record => {
+    const parentEntityUuid = R.prop(NodeKeys.keys.nodeUuid, R.last(record.keysHierarchy))
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportView.js]: I would create a method in NodeKeys to get keysHierarchy. it's not a good practice to access properties directly

File: webapp/commonComponents/validationFieldMessages.js [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+      <div className={`validation-field_message ${type}`} key={i}>
+        { showIcons && <span className="icon icon-warning icon-12px icon-left" /> }
+        <Markdown className="validation-field-message__text" source={message}/>
+      </div>
+    )
+  )(validation)
+}
+
+const mapStateToProps = state => {
+  return { survey: SurveyState.getSurvey(state) }
```
Comment (minotogna) [inline webapp/commonComponents/validationFieldMessages.js]: I wouldn't pass the entire survey to the component

File: webapp/commonComponents/validationFieldMessages.js [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+
+  R.pipe(
+    getValidationErrorMessages(i18n),
+    x => { if (x.length) messages.push(x) }
+  )(validation)
+
+  return messages
+}
+
+const ValidationFieldMessages = ({ validation, showKeys = true, showIcons = false, survey }) => {
```
Comment (minotogna) [inline webapp/commonComponents/validationFieldMessages.js]: it's more readable if you use props.

File: webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js [hunk 25]
```diff
@@ -63,8 +63,9 @@ const mapStateToProps = (state, props) => {
       if (node) {
         validation = RecordValidation.getNodeValidation(node)(recordValidation)
       } else if (NodeDef.isEntity(nodeDef)) {
-        //only entities can have children with min/max count validation
-        validation = RecordValidation.getValidationChildrenCount(parentNode, nodeDef)(recordValidation)
+        // only entities can have children with min/max count validation
+        const fieldValidation = Validation.getFieldValidation(NodeDef.getUuid(parentNode))(recordValidation)
```
[the remark this answers is not in the document]
Comment (sgiaccio) [reply webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefErrorBadge.js]: Right, changed

File: webapp/loggedin/modules/data/validationReport/validationReportRow.js (hunk 21 above)
Comment (SteRiccio) [reply webapp/loggedin/modules/data/validationReport/validationReportRow.js]: done

File: webapp/loggedin/modules/data/validationReport/validationReportView.js (hunk 22 above)
Comment (SteRiccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.js]: refactored almost entirely

File: webapp/commonComponents/validationFieldMessages.js (hunk 24 above)
Comment (SteRiccio) [reply webapp/commonComponents/validationFieldMessages.js]: right, done

File: webapp/commonComponents/validationFieldMessages.js (hunk 23 above)
Comment (SteRiccio) [reply webapp/commonComponents/validationFieldMessages.js]: yes... me neither.. but how can we do it?

File: webapp/loggedin/modules/data/dataVis/dataQuery/components/tableRows.js [hunk 26]
```diff
@@ -37,7 +37,7 @@ const TableRows = ({ nodeDefCols, data, offset, lang, colWidth, defaultColWidth,
             const { parentNodeUuid, record } = row
 
             const recordUuid = Record.getUuid(record)
-            const recordEditUrl = `${appModuleUri(dataModules.record)}${recordUuid}?parentNodeUuid=${parentNodeUuid}`
+            const recordEditUrl = `${appModuleUri(dataModules.record)}${recordUuid}?focusNodeUuid=${parentNodeUuid}`
```
Comment (minotogna) [inline webapp/loggedin/modules/data/dataVis/dataQuery/components/tableRows.js]: I would call the param ```pageNodeUuid```

File: webapp/loggedin/modules/data/validationReport/validationReportView.js [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+    reloadListItems(validationReportModule, { cycle: surveyCycleKey })
+  }, [surveyCycleKey])
+
+  const onRowClick = row => {
+    const focusNodeUuid = RecordValidationReportItem.getNodeContextUuid(row)
+    const focusNodeDefUuid = RecordValidationReportItem.getNodeDefContextUuid(row)
+    const recordUuid = RecordValidationReportItem.getRecordUuid(row)
+    const recordEditUrl = `${appModuleUri(
+      dataModules.record,
+    )}${recordUuid}?focusNodeUuid=${focusNodeUuid}&focusNodeDefUuid=${focusNodeDefUuid}`
```
Comment (minotogna) [inline webapp/loggedin/modules/data/validationReport/validationReportView.js]: i would call the params the ```page``` prefix . e.g. ```pageNodeDefUuid```

File: webapp/commonComponents/validationFieldMessages.js [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+
+ValidationFieldMessages.defaultProps = {
+  validation: null,
+  survey: null,
+  showKeys: true, // Show the key of the fields' validations
+  showIcons: false, // Show error or warning icon
+}
+
+const mapStateToProps = state => ({
+  survey: SurveyState.getSurvey(state),
```
Comment (minotogna) [inline webapp/commonComponents/validationFieldMessages.js]: I wouldn't pass the survey to the component. can u use getValidationFieldMessages in mapStateToProps?

File: core/i18n/resources/en.js [hunk 29]
```diff
@@ -41,12 +41,14 @@ export const enTranslation = {
     label_plural: 'Labels',
     language: 'Language',
     manage: 'Manage',
+    messages: 'Messages',
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply core/i18n/resources/en.js]: right, done

File: webapp/loggedin/modules/data/dataVis/dataQuery/components/tableRows.js (hunk 26 above)
Comment (SteRiccio) [reply webapp/loggedin/modules/data/dataVis/dataQuery/components/tableRows.js]: ok

File: webapp/loggedin/modules/data/validationReport/validationReportView.js (hunk 27 above)
Comment (SteRiccio) [reply webapp/loggedin/modules/data/validationReport/validationReportView.js]: ok

File: webapp/commonComponents/validationFieldMessages.js (hunk 28 above)
Comment (SteRiccio) [reply webapp/commonComponents/validationFieldMessages.js]: Of course!

**Pull request merged**